### DISCUSSION TITLE
Expression declared locals

### DISCUSSION THREAD

**Discussion by TahirAhmadov** [dotnet/csharplang#5404] [Language Ideas] [answered]
We already have this:
```cs
if(o is string s) { ... }
```
Can we also get something like this:
```cs
if((string? s = GetString()) != null) { ... }
```
This is useful when there are many `if` statements:
```cs
if((string? s = GetString()) != null) { ... }
else if((int? i = GetInt()) != null) { ... }
// as opposed to
string? s;
int? i;
// ... many more variables ...
if((s = GetString()) != null) { ... }
else if((i = GetInt()) != null) { ... }
// ... many more else-if statements ...
```

**Comment by 333fred** [comment]
What advantage does this have over pattern matching? 

```cs
if (GetString() is string s)
```

**Reply by TahirAhmadov** [reply]
It would allow many other conditions - the `!= null` was just an example. However, having thought about it, I just realized I can use a "noop pattern match" and append `&& condition` like so: `if(GetInt() is int i && i > 5) { ... }`. I started this discussion specifically because that's such an awkward syntax to accomplish the same result, it didn't even occur to me I can use it for this purpose. I understand it's a low priority, but it would be nice to declare locals without using "noop pattern matches".

**Reply by svick** [reply]
@TahirAhmadov Personally, I find parenthesized variable declaration to be quite awkward syntax. Using always-matching patterns is unappealing in a different way, but I'm getting used it. Note that you can also simplify that code a tiny bit to `GetInt() is var i and > 5`.

**Comment by TahirAhmadov** [comment]
I would like to bump this discussion. The pattern matching solution can be used if the whole expression is a condition - like `if` statement - but it doesn't work in all other, non-conditional cases. Currently I have to do this:
```cs
public ClassCtor()
{
  this.Controls.Add(new Div
  {
    (this._lblName = new Label { Text = "Name" }),
    new TextBox { Id = "tbName" }.Init(x => this._lblName.For = x),
  });
}
// ....
readonly Label _lblName;
```
The problem is that now the class is polluted with useless fields for controls which are only accessed inside the ctor once. Another approach could be:
```cs
public ClassCtor()
{
  Label lblName;
  this.Controls.Add(new Div
  {
    (lblName = new Label { Text = "Name" }),
    new TextBox { Id = "tbName" }.Init(x => lblName.For = x),
  });
}
```
Which would be fine, but now the ctor becomes (even) longer and unreadable - if we have a few dozen of such controls. The real solution would be this:
```cs
public ClassCtor()
{
  this.Controls.Add(new Div
  {
    (var lblName = new Label { Text = "Name" }),
    new TextBox { Id = "tbName" }.Init(x => lblName.For = x),
  });
}
```

**Reply by Yarith** [reply]
For this exact use case i have this extension method:
```csharp
static class Extensions
{
    public static T Let<T>(this T instance, out T @this) =>
        @this = instance;
}
```

You should then be able to use it like this:
```csharp
this.Controls.Add(new Div
{
    new Label { Text = "Name" }.Let(out var lblName),
    new TextBox { Id = "tbName" }.Init(x => lblName.For = x),
});
```

**Reply by TahirAhmadov** [reply]
@Yarith thank you, this is helpful, I hadn't thought of that.

**Comment by 333fred** [comment, marked as answer]
Duplicate of https://github.com/dotnet/csharplang/issues/973.
